src/huggingface_hub/inference/_mcp/mcp_client.py

```diff
@@ -233,6 +248,28 @@ async def add_mcp_server(self, type: ServerType, **params: Any):
                 )
             )
 
+    def _filter_tools(self, tools: List[Any], allowed_tools: Optional[List[str]]) -> List[Any]:
+        """Filter tools based on allowed_tools list.
+
+        Args:
+            tools: List of MCP tool objects
+            allowed_tools: Optional list of tool names to allow
+
+        Returns:
+            Filtered list of tools
+        """
+        if allowed_tools is None:
+            return tools
+
+        # Validate that specified tools exist
+        all_tool_names = [tool.name for tool in tools]
+        missing_tools = set(allowed_tools) - set(all_tool_names)
+        if missing_tools:
+            logger.warning(f"Tools specified in 'allowed_tools' not found on server: {list(missing_tools)}")
+
+        # Filter tools using list comprehension
+        return [tool for tool in tools if tool.name in allowed_tools]
```

Comment:
This method is far too complex and verbose IMO. Would be better to remove it and replace by a simple `[tool for tool in tools if tool.name in allowed_tools]`.